src/lib/yacc/parser.rs

```diff
@@ -0,0 +1,751 @@
+        Regex::new("^(?:(\".+?\")|('.+?')|([a-zA-Z_][a-zA-Z_0-9]*))").unwrap()
+    };
+}
+
+/// The actual parser is intended to be entirely opaque from outside users.
+impl YaccParser {
+    fn new(src: String) -> YaccParser {
+        YaccParser {
+            src     : src,
+            newlines: vec![0],
+            grammar : GrammarAST::new()
+        }
+    }
+
+    fn mk_error(&self, k: YaccParserErrorKind, off: usize) -> YaccParserError {
+        let (line, col) = self.off_to_line_col(off);
+        YaccParserError{kind: k, line: line, col: col}
+    }
+
+    fn off_to_line_col(&self, off: usize) -> (usize, usize) {
+        if off == self.src.len() {
+            let line_off = *self.newlines.iter().last().unwrap();
+            return (self.newlines.len(), self.src[line_off..].chars().count() + 1);
+        }
+        let (line_m1, &line_off) = self.newlines.iter()
+                                                .enumerate()
+                                                .rev()
+                                                .find(|&(_, &line_off)| line_off <= off)
+                                                .unwrap();
+        let c_off = self.src[line_off..]
+                        .char_indices()
+                        .position(|(c_off, _)| c_off == off - line_off)
+                        .unwrap();
+        (line_m1 + 1, c_off + 1)
+    }
+
+    fn parse(&mut self) -> YaccResult<usize> {
+        // We pass around an index into the *bytes* of self.src. We guarantee that at all times
+        // this points to the beginning of a UTF-8 character (since multibyte characters exist, not
+        // every byte within the string is also a valid character).
+        let mut i = try!(self.parse_declarations(0));
+        i = try!(self.parse_rules(i));
+        // We don't currently support the programs part of a specification. One day we might...
+        match self.lookahead_is("%%", i) {
+            Some(j) => {
+                if try!(self.parse_ws(j)) == self.src.len() { Ok(i) }
+                else {
+                    Err(self.mk_error(YaccParserErrorKind::ProgramsNotSupported, i))
+                }
+            }
+            None    => Ok(i)
+        }
+    }
+
+    fn parse_declarations(&mut self, mut i: usize) -> YaccResult<usize> {
+        i = try!(self.parse_ws(i));
+        let mut prec_level  = 0;
+        while i < self.src.len() {
+            if self.lookahead_is("%%", i).is_some() { return Ok(i); }
+            if let Some(j) = self.lookahead_is("%token", i) {
+                i = try!(self.parse_ws(j));
+                while i < self.src.len() {
+                    if self.lookahead_is("%", i).is_some() { break; }
+                    let (j, n) = try!(self.parse_terminal(i));
+                    self.grammar.tokens.insert(n);
+                    i = try!(self.parse_ws(j));
+                }
+            }
+            else if let Some(j) = self.lookahead_is("%start", i) {
```

Comment:
Sometimes here you say:
```rust
if ... {
}
else {
}
```
and sometimes:
```rust
if ... {
} else {
}
```
Maybe better to stick with one convention?